### PR TITLE
sql: add DROP {CONSTRAINT,COLUMN} support

### PR DESCRIPTION
**Pull request by tamird**
[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/2929)

### CODE REVIEW

**Comment by vivekmenezes** [comment]
hey can you wait until my PR is in? It's going to take a day or two.

**Comment by petermattis** [comment]
We need to do a better job of coordinating work. Someone is going to have an ugly merge. 


---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1emcIfXHdfxaOzjiad)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L103)):</sup>
Add a separating blank line before the case statement for consistency.


---

<sup>**[sql/alter_table.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1enES43iAcJ-0apDW4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L114)):</sup>
`s/is contained in/is referenced by/g`? Not sure about this.


---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
This can't happen, right? Similar to another comment, perhaps `FindColumnByName` should return the position of the column.


---

<sup>**[sql/alter_table.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1empg31FV3-iwSQ9-L)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L153)):</sup>
This can't happen, right? Perhaps `FindIndexByName` could return the position of the index.


---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
I think this is only correct because you know how columns are being removed which feels a bit fragile. I'd prefer passing in the list of dropped columns explicitly, just as you used to do with the add indexes. 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by tamird** [comment]
Not it!


---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1emcIfXHdfxaOzjiad)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L103)):</sup>
Done.


---

<sup>**[sql/alter_table.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1enES43iAcJ-0apDW4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L114)):</sup>
Done.


---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
Done.


---

<sup>**[sql/alter_table.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1empg31FV3-iwSQ9-L)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L153)):</sup>
Done.


---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
I think the assumption that columns are not reordered is a valid one, don't you?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by petermattis** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
What was done? I'm not seeing any changes here.


---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
It's not called out anywhere. And since this is only called from one place, it doesn't seem any easier to regenerate the list of dropped columns vs keeping track of it as they are dropped.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by vivekmenezes** [comment]
Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
do you want to panic if j >= len(newTableDesc.Columns)? It's probably better to make this code more self contained.


---

<sup>**[sql/backfill.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ioGBIXgmk66MHQN6Q)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L43)):</sup>
Add a comment with some sql examples of what this is doing.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by petermattis** [comment]
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ipQhzj8crffP10FZp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L48)):</sup>
Does this work if you're dropping more than one column? I would think you would need something like `UPDATE <table> SET (a, b) = (NULL, NULL)`


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by tamird** [comment]
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[sql/alter_table.go, line 131 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1en4vENLuZqvEjOCjQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/alter_table.go#L131)):</sup>
Oh, this one wasn't done. Done now.


---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
Peter: it's called from two places. Vivek: Fixed this to be safe in that case.


---

<sup>**[sql/backfill.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ioGBIXgmk66MHQN6Q)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L43)):</sup>
Done.


---

<sup>**[sql/backfill.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1ipQhzj8crffP10FZp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/a448d8ea3e6d820d3b0d7e3aa5e0588d65f92fa7/sql/backfill.go#L48)):</sup>
Fixed and added a test.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by petermattis** [comment]
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
It's called from two places, but only one drops column descriptors.


---

<sup>**[sql/backfill.go, line 26 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9MkIpnRHaKDDHS1D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L26)):</sup>
`s/makecol/makeCol/g`


---

<sup>**[sql/backfill.go, line 59 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9rOH9tdm3QB16w0U)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L59)):</sup>
I believe this is actually `SET col1 = NULL, col2 = NULL, ...`


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by tamird** [comment]
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
Again, that's true, but I really don't want to push that responsibility to the caller. I've added a comment documenting the assumption here.


---

<sup>**[sql/backfill.go, line 26 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9MkIpnRHaKDDHS1D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L26)):</sup>
Done.


---

<sup>**[sql/backfill.go, line 59 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1j9rOH9tdm3QB16w0U)** ([raw file](https://github.com/cockroachdb/cockroach/blob/78c401c90dba9743168b6aa6bfa7c83d0fc1695c/sql/backfill.go#L59)):</sup>
That's true, updated.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

File: sql/backfill.go [hunk 1]
```diff
@@ -23,28 +23,75 @@ import (
 	"github.com/cockroachdb/cockroach/util/log"
 )
 
-func (p *planner) makeBackfillBatch(tableName *parser.QualifiedName, tableDesc *TableDescriptor, indexDescs ...IndexDescriptor) (client.Batch, error) {
-	b := client.Batch{}
+func makeColIDtoRowIndex(row planNode, desc *TableDescriptor) (map[ColumnID]int, error) {
+	colIDtoRowIndex := make(map[ColumnID]int)
```
Comment (vivekmenezes) [inline sql/backfill.go]: make(, len(row.Columns())) for size of the map

Comment (tamird) [inline sql/backfill.go]: done

**Comment by vivekmenezes** [comment]
LGTM

**Comment by tamird** [comment]
Review status: 0 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[sql/backfill.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1engH1q56Mbc-0QatL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a8adebc69eac6b7176cf00fede511429b296d4/sql/backfill.go#L31)):</sup>
I've changed this to sort the descriptors before attempting to work out the ones that were removed.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by petermattis** [comment]
LGTM


---

Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

<sup>**[sql/backfill.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jgqlx2sGTe3pkLM_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L45)):</sup>
We generally don't put type terminology such as `Slice` in the type name. I would have named this `columnsByID`.


---

<sup>**[sql/backfill.go, line 57 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jh0wA6wvH60YOitMj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L57)):</sup>
`indexesByID`?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by tamird** [comment]
Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

<sup>**[sql/backfill.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jgqlx2sGTe3pkLM_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L45)):</sup>
Done.


---

<sup>**[sql/backfill.go, line 57 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929#-K1jh0wA6wvH60YOitMj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/191b6dfbdde9cf39908a1daf36cf642be4c4f7c1/sql/backfill.go#L57)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Comment by petermattis** [comment]
LGTM, just be sure to coordinate merging with Vivek.


---

Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2929)

**Pull request merged**